From b7a5af3b1abb656ddd83586247a2fd9734a361b8 Mon Sep 17 00:00:00 2001
From: kiddin9 <48883331+kiddin9@users.noreply.github.com>
Date: Sun, 14 Apr 2024 14:03:26 +0800
Subject: [PATCH 1/2] Update luci-app-filebrowser.patch Update .gitignore
 Update upstream.yml Update upstream.yml Update upstream.yml Update
 upstream.yml Update 400-fix.patch Update 400-fix.patch Create miniupnpd.patch
 Update upnpd.config Delete
 .github/diy/packages/miniupnpd-iptables/patches/106-spam-syslog-ignoring.patch
 Create 400-fix.patch Delete
 .github/diy/packages/miniupnpd-iptables/Makefile.k Update upstream.yml Update
 upstream.yml Update upstream.yml Update .gitignore Update upstream.yml Update
 advancedplus.patch Update 040-wireless-5.8.patch Update
 106-spam-syslog-ignoring.patch Update .gitignore Update
 106-spam-syslog-ignoring.patch Create 106-spam-syslog-ignoring.patch Update
 Makefile Update upstream.yml update Update 030-wireless-5.8.patch Update
 040-wireless-5.8.patch Update 030-wireless-5.8.patch Update
 040-wireless-5.8.patch Create 040-wireless-5.8.patch Create
 030-wireless-5.8.patch Create 020-wireless-5.8.patch Delete
 .github/diy/patches/rtl8189es/patches/020-wireless-5.8.patch Update
 upstream.yml Update luci-app-turboacc.patch update Update upstream.yml

---
 .../packages/miniupnpd-iptables/Makefile.k    |  83 ----
 .../packages/miniupnpd/patches/400-fix.patch  | 184 +++++++
 .../rtl8188eu/patches/030-wireless-5.8.patch  | 113 +++++
 .../rtl8189es/patches/020-wireless-5.8.patch  |  86 ++++
 .../rtl8192eu/patches/030-wireless-5.8.patch  | 112 +++++
 .../patches/040-wireless-5.8.patch            | 139 ++++++
 .github/diy/patches/advancedplus.patch        |  51 +-
 .../diy/patches/luci-app-filebrowser.patch    |   4 +-
 .github/diy/patches/luci-app-turboacc.patch   |   2 +-
 .github/diy/patches/miniupnpd.patch           |  13 +
 .github/workflows/upstream.yml                |  45 +-
 .gitignore                                    |   1 +
 miniupnpd/files/upnpd.config                  |   4 +
 my-default-settings/Makefile                  |   2 +-
 my-default-settings/files/sbin/sysupgrade     | 375 ---------------
 .../files/NanoHatOLED/DejaVuSansMono-Bold.ttf | Bin 229460 -> 0 bytes
 .../files/NanoHatOLED/DejaVuSansMono.ttf      | Bin 237788 -> 0 bytes
 .../files/NanoHatOLED/bakebit_128_64_oled.py  | 455 ------------------
 .../files/NanoHatOLED/bakebit_nanohat_oled.py | 345 -------------
 nanohatoled/files/NanoHatOLED/logo.png        | Bin 1421 -> 0 bytes
 nanohatoled/files/nanohatoled                 |  20 -
 rtl8188eu/patches/030-wireless-5.8.patch      |  18 +
 22 files changed, 713 insertions(+), 1339 deletions(-)
 delete mode 100644 .github/diy/packages/miniupnpd-iptables/Makefile.k
 create mode 100644 .github/diy/packages/miniupnpd/patches/400-fix.patch
 create mode 100644 .github/diy/packages/rtl8188eu/patches/030-wireless-5.8.patch
 create mode 100644 .github/diy/packages/rtl8189es/patches/020-wireless-5.8.patch
 create mode 100644 .github/diy/packages/rtl8192eu/patches/030-wireless-5.8.patch
 create mode 100644 .github/diy/packages/rtl8812au-ac/patches/040-wireless-5.8.patch
 create mode 100644 .github/diy/patches/miniupnpd.patch
 delete mode 100755 my-default-settings/files/sbin/sysupgrade
 delete mode 100644 nanohatoled/files/NanoHatOLED/DejaVuSansMono-Bold.ttf
 delete mode 100644 nanohatoled/files/NanoHatOLED/DejaVuSansMono.ttf
 delete mode 100644 nanohatoled/files/NanoHatOLED/bakebit_128_64_oled.py
 delete mode 100644 nanohatoled/files/NanoHatOLED/bakebit_nanohat_oled.py
 delete mode 100644 nanohatoled/files/NanoHatOLED/logo.png
 delete mode 100755 nanohatoled/files/nanohatoled

diff --git a/.github/diy/packages/miniupnpd-iptables/Makefile.k b/.github/diy/packages/miniupnpd-iptables/Makefile.k
deleted file mode 100644
index 80c50fa37..000000000
--- a/.github/diy/packages/miniupnpd-iptables/Makefile.k
+++ /dev/null
@@ -1,83 +0,0 @@
-#
-# Copyright (C) 2006-2014 OpenWrt.org
-#
-# This is free software, licensed under the GNU General Public License v2.
-# See /LICENSE for more information.
-#
-
-include $(TOPDIR)/rules.mk
-
-PKG_NAME:=miniupnpd-iptables
-PKG_VERSION:=3022-08-31
-PKG_RELEASE:=25
-
-PKG_SOURCE_URL:=http://miniupnp.free.fr/files
-PKG_SOURCE:=miniupnpd-2.0.20180503.tar.gz
-PKG_BUILD_DIR:=$(BUILD_DIR)/miniupnpd-2.0.20180503
-PKG_HASH:=skip
-
-PKG_MAINTAINER:=Markus Stenberg <fingon@iki.fi>
-PKG_LICENSE:=BSD-3-Clause
-
-include $(INCLUDE_DIR)/package.mk
-
-define Package/$(PKG_NAME)
-  SECTION:=net
-  CATEGORY:=Network
-  DEPENDS:=+iptables +libip4tc +libuuid
-  PROVIDES:=miniupnpd
-  TITLE:=Lightweight UPnP IGD, NAT-PMP & PCP daemon
-  SUBMENU:=Firewall
-  URL:=http://miniupnp.free.fr/
-  VARIANT:=iptables
-endef
-
-define Package/$(PKG_NAME)/config
-config MINIUPNPD_IGDv2
-	bool
-	default n
-	prompt "Enable IGDv2"
-endef
-
-define Package/$(PKG_NAME)/conffiles
-/etc/config/upnpd
-endef
-
-define Package/$(PKG_NAME)/postinst
-#!/bin/sh
-
-if [ -z "$$IPKG_INSTROOT" ]; then
-  ( . /etc/uci-defaults/99-miniupnpd )
-  rm -f /etc/uci-defaults/99-miniupnpd
-fi
-
-exit 0
-endef
-
-define Build/Prepare
-	$(call Build/Prepare/Default)
-	echo "OpenWrt" | tr \(\)\  _ >$(PKG_BUILD_DIR)/os.openwrt
-endef
-
-MAKE_FLAGS += \
-	TARGET_OPENWRT=1 TEST=0 \
-	LIBS="" \
-	CC="$(TARGET_CC) -DIPTABLES_143 \
-		-lip4tc -luuid" \
-	CONFIG_OPTIONS="--portinuse --leasefile \
-		$(if $(CONFIG_MINIUPNPD_IGDv2),--igd2)" \
-	-f Makefile.linux \
-	miniupnpd
-
-
-define Package/$(PKG_NAME)/install
-	$(INSTALL_DIR) $(1)/usr/sbin $(1)/etc/init.d $(1)/etc/config $(1)/etc/uci-defaults $(1)/etc/hotplug.d/iface $(1)/usr/share/miniupnpd
-	$(INSTALL_BIN) $(PKG_BUILD_DIR)/miniupnpd $(1)/usr/sbin/miniupnpd
-	$(INSTALL_BIN) ./files/miniupnpd.init $(1)/etc/init.d/miniupnpd
-	$(INSTALL_CONF) ./files/upnpd.config $(1)/etc/config/upnpd
-	$(INSTALL_DATA) ./files/miniupnpd.hotplug $(1)/etc/hotplug.d/iface/50-miniupnpd
-	$(INSTALL_DATA) ./files/miniupnpd.defaults $(1)/etc/uci-defaults/99-miniupnpd
-	$(INSTALL_DATA) ./files/firewall.include $(1)/usr/share/miniupnpd/firewall.include
-endef
-
-$(eval $(call BuildPackage,$(PKG_NAME)))
diff --git a/.github/diy/packages/miniupnpd/patches/400-fix.patch b/.github/diy/packages/miniupnpd/patches/400-fix.patch
new file mode 100644
index 000000000..5724dbee0
--- /dev/null
+++ b/.github/diy/packages/miniupnpd/patches/400-fix.patch
@@ -0,0 +1,184 @@
+--- a/getifaddr.c
++++ b/getifaddr.c
+@@ -309,7 +309,7 @@ addr_is_reserved(struct in_addr * addr)
+ 
+ 	for (i = 0; i < sizeof(reserved)/sizeof(reserved[0]); ++i) {
+ 		if ((address >> reserved[i].rmask) == (reserved[i].address >> reserved[i].rmask))
+-			return 1;
++			return 0;
+ 	}
+ 
+ 	return 0;
+
+--- a/miniupnpd.c
++++ b/miniupnpd.c
+@@ -1014,7 +1014,7 @@ parselanaddr(struct lan_addr_s * lan_addr, const char * str, int debug_flag)
+ 				INIT_PRINT_ERR("Error parsing address : %s\n", lan_addr->ext_ip_str);
+ 				return -1;
+ 			}
+-			if(addr_is_reserved(&lan_addr->ext_ip_addr)) {
++			if(addr_is_reserved(&lan_addr->ext_ip_addr) && !GETFLAG(FORCEFORWARDINGMASK)) {
+ 				/* error */
+ 				INIT_PRINT_ERR("Error: option ext_ip address contains reserved / private address : %s\n", lan_addr->ext_ip_str);
+ 				return -1;
+@@ -1256,6 +1256,10 @@ init(int argc, char * * argv, struct runtime_vars * v)
+ 			case UPNPEXT_IP:
+ 				use_ext_ip_addr = ary_options[i].value;
+ 				break;
++			case UPNP_FORCE_FORWARDING:
++				if(strcmp(ary_options[i].value, "yes") == 0)
++					SETFLAG(FORCEFORWARDINGMASK);
++				break;
+ 			case UPNPEXT_PERFORM_STUN:
+ 				if(strcmp(ary_options[i].value, "yes") == 0)
+ 					SETFLAG(PERFORMSTUNMASK);
+@@ -1858,7 +1862,7 @@ init(int argc, char * * argv, struct runtime_vars * v)
+ 			INIT_PRINT_ERR("Error: option ext_ip contains invalid address %s\n", use_ext_ip_addr);
+ 			return 1;
+ 		}
+-		if (addr_is_reserved(&addr)) {
++		if (addr_is_reserved(&addr) && !GETFLAG(FORCEFORWARDINGMASK)) {
+ 			INIT_PRINT_ERR("Error: option ext_ip contains reserved / private address %s, not public routable\n", use_ext_ip_addr);
+ 			return 1;
+ 		}
+@@ -2304,7 +2308,7 @@ main(int argc, char * * argv)
+ 		if (getifaddr(ext_if_name, if_addr, INET_ADDRSTRLEN, &addr, NULL) < 0) {
+ 			syslog(LOG_WARNING, "Cannot get IP address for ext interface %s. Network is down", ext_if_name);
+ 			disable_port_forwarding = 1;
+-		} else if (addr_is_reserved(&addr)) {
++		} else if (addr_is_reserved(&addr) && !GETFLAG(FORCEFORWARDINGMASK)) {
+ 			syslog(LOG_INFO, "Reserved / private IP address %s on ext interface %s: Port forwarding is impossible", if_addr, ext_if_name);
+ 			syslog(LOG_INFO, "You are probably behind NAT, enable option ext_perform_stun=yes to detect public IP address");
+ 			syslog(LOG_INFO, "Or use ext_ip= / -o option to declare public IP address");
+@@ -2595,7 +2599,7 @@ main(int argc, char * * argv)
+ 					syslog(LOG_WARNING, "Cannot get IP address for ext interface %s. Network is down", ext_if_name);
+ 					disable_port_forwarding = 1;
+ 				} else {
+-					int reserved = addr_is_reserved(&addr);
++					int reserved = addr_is_reserved(&addr) && !GETFLAG(FORCEFORWARDINGMASK);
+ 					if (!disable_port_forwarding && reserved) {
+ 						syslog(LOG_INFO, "Reserved / private IP address %s on ext interface %s: Port forwarding is impossible", if_addr, ext_if_name);
+ 						syslog(LOG_INFO, "You are probably behind NAT, enable option ext_perform_stun=yes to detect public IP address");
+
+--- a/miniupnpd.conf
++++ b/miniupnpd.conf
+@@ -9,6 +9,9 @@
+ # Setting ext_ip is also useful in double NAT setup, you can declare here
+ # the public IP address.
+ #ext_ip=
++
++#force forwarding enable for upnp: default is no
++#force_forwarding=yes
+ # WAN interface must have public IP address. Otherwise it is behind NAT
+ # and port forwarding is impossible. In some cases WAN interface can be
+ # behind unrestricted full-cone NAT 1:1 when all incoming traffic is NAT-ed and
+
+--- a/natpmp.c
++++ b/natpmp.c
+@@ -109,7 +109,7 @@ static void FillPublicAddressResponse(unsigned char * resp, in_addr_t senderaddr
+ 			syslog(LOG_ERR, "Failed to get IP for interface %s", ext_if_name);
+ 			resp[3] = 3;	/* Network Failure (e.g. NAT box itself
+ 			                 * has not obtained a DHCP lease) */
+-		} else if (addr_is_reserved(&addr)) {
++		} else if (addr_is_reserved(&addr) && !GETFLAG(FORCEFORWARDINGMASK)) {
+ 			resp[3] = 3;	/* Network Failure, box has not obtained
+ 			                   public IP address */
+ 		} else {
+
+--- a/options.c
++++ b/options.c
+@@ -35,6 +35,7 @@ static const struct {
+ 	{ UPNPEXT_IFNAME6, "ext_ifname6" },
+ #endif
+ 	{ UPNPEXT_IP,	"ext_ip" },
++	{ UPNP_FORCE_FORWARDING, "force_forwarding" },
+ 	{ UPNPEXT_PERFORM_STUN, "ext_perform_stun" },
+ 	{ UPNPEXT_STUN_HOST, "ext_stun_host" },
+ 	{ UPNPEXT_STUN_PORT, "ext_stun_port" },
+
+--- a/options.h
++++ b/options.h
+@@ -21,6 +21,7 @@ enum upnpconfigoptions {
+ 	UPNPEXT_IFNAME6,		/* ext_ifname6 */
+ #endif
+ 	UPNPEXT_IP,				/* ext_ip */
++	UPNP_FORCE_FORWARDING, /* force forwarding enable for upnp */
+ 	UPNPEXT_PERFORM_STUN,		/* ext_perform_stun */
+ 	UPNPEXT_STUN_HOST,		/* ext_stun_host */
+ 	UPNPEXT_STUN_PORT,		/* ext_stun_port */
+
+--- a/testgetifaddr.c
++++ b/testgetifaddr.c
+@@ -13,6 +13,8 @@
+ #include "config.h"
+ #include "getifaddr.h"
+ 
++int runtime_flags = 0;
++
+ #if defined(__sun)
+ /* solaris 10 does not define LOG_PERROR */
+ #define LOG_PERROR 0
+
+--- a/testportinuse.c
++++ b/testportinuse.c
+@@ -14,6 +14,8 @@
+ #include "config.h"
+ #include "portinuse.h"
+ 
++int runtime_flags = 0;
++
+ int main(int argc, char * * argv)
+ {
+ #ifndef CHECK_PORTINUSE
+
+--- a/upnpdescgen.c
++++ b/upnpdescgen.c
+@@ -1316,7 +1316,7 @@ genEventVars(int * len, const struct serviceDesc * s)
+ 				else {
+ 					struct in_addr addr;
+ 					char ext_ip_addr[INET_ADDRSTRLEN];
+-					if(getifaddr(ext_if_name, ext_ip_addr, INET_ADDRSTRLEN, &addr, NULL) < 0 || addr_is_reserved(&addr)) {
++					if(getifaddr(ext_if_name, ext_ip_addr, INET_ADDRSTRLEN, &addr, NULL) < 0 || (addr_is_reserved(&addr) && !GETFLAG(FORCEFORWARDINGMASK))) {
+ 						str = strcat_str(str, len, &tmplen, "0.0.0.0");
+ 					} else {
+ 						str = strcat_str(str, len, &tmplen, ext_ip_addr);
+
+--- a/upnpglobalvars.h
++++ b/upnpglobalvars.h
+@@ -87,6 +87,8 @@ extern int runtime_flags;
+ 
+ #define PERFORMSTUNMASK    0x1000
+ 
++#define FORCEFORWARDINGMASK 0x2000
++
+ #define SETFLAG(mask)	runtime_flags |= mask
+ #define GETFLAG(mask)	(runtime_flags & mask)
+ #define CLEARFLAG(mask)	runtime_flags &= ~mask
+
+--- a/upnpredirect.c
++++ b/upnpredirect.c
+@@ -444,7 +444,7 @@ upnp_redirect_internal(const char * rhost, unsigned short eport,
+ {
+ 	/*syslog(LOG_INFO, "redirecting port %hu to %s:%hu protocol %s for: %s",
+ 		eport, iaddr, iport, protocol, desc);			*/
+-	if(disable_port_forwarding)
++	if(disable_port_forwarding && !GETFLAG(FORCEFORWARDINGMASK))
+ 		return -1;
+ 	if(add_redirect_rule2(ext_if_name, rhost, eport, iaddr, iport, proto,
+ 	                      desc, timestamp) < 0) {
+
+--- a/upnpsoap.c
++++ b/upnpsoap.c
+@@ -348,7 +348,11 @@ GetExternalIPAddress(struct upnphttp * h, const char * action, const char * ns)
+ 			ext_ip_addr[0] = '\0';
+ 		} else if (addr_is_reserved(&addr)) {
+ 			syslog(LOG_NOTICE, "private/reserved address %s is not suitable for external IP", ext_ip_addr);
+-			ext_ip_addr[0] = '\0';
++			if (!GETFLAG(FORCEFORWARDINGMASK)) {
++				ext_ip_addr[0] = '\0';
++			} else {
++				syslog(LOG_NOTICE, "force_forwarding enable, private/reserved address %s used as external IP", ext_ip_addr);
++			}
+ 		}
+ 	}
+ #else
diff --git a/.github/diy/packages/rtl8188eu/patches/030-wireless-5.8.patch b/.github/diy/packages/rtl8188eu/patches/030-wireless-5.8.patch
new file mode 100644
index 000000000..8556f682d
--- /dev/null
+++ b/.github/diy/packages/rtl8188eu/patches/030-wireless-5.8.patch
@@ -0,0 +1,113 @@
+--- a/include/ieee80211.h
++++ b/include/ieee80211.h
+@@ -1529,7 +1529,7 @@ enum ieee80211_state {
+ 	(((Addr[2]) & 0xff) == 0xff) && (((Addr[3]) & 0xff) == 0xff) && (((Addr[4]) & 0xff) == 0xff) && \
+ 				     (((Addr[5]) & 0xff) == 0xff))
+ #else
+-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 00))
++#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 18, 00) && !defined(BUILD_OPENWRT))
+ extern __inline int is_multicast_mac_addr(const u8 *addr)
+ {
+ 	return (addr[0] != 0xff) && (0x01 & addr[0]);
+--- a/os_dep/linux/ioctl_cfg80211.c
++++ b/os_dep/linux/ioctl_cfg80211.c
+@@ -417,9 +417,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset, u8
+ 	if (ret != _SUCCESS)
+ 		goto exit;
+ 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+         cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0);
+ #else
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef);
+@@ -1105,7 +1105,7 @@ void rtw_cfg80211_indicate_connect(_adapter *padapter)
+ 		#endif
+ 
+ 		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 12, 0) || defined(RHEL79))
+-		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
++		#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0) || defined(BUILD_OPENWRT))
+                 roam_info.links[0].channel = notify_channel;
+ 		roam_info.links[0].bssid = cur_network->network.MacAddress;
+ 		#else
+@@ -1667,7 +1667,7 @@ static int rtw_cfg80211_set_encryption(struct net_device *dev, struct ieee_param
+ }
+ 
+ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         , int link_id
+ #endif
+ 	, u8 key_index
+@@ -1814,7 +1814,7 @@ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+ }
+ 
+ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         , int link_id
+ #endif
+ 	, u8 keyid
+@@ -1982,7 +1982,7 @@ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+ }
+ 
+ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         int link_id,
+ #endif
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
+@@ -2006,7 +2006,7 @@ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+ 
+ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         int link_id,
+ #endif
+         u8 key_index
+@@ -2058,7 +2058,7 @@ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 30))
+ int cfg80211_rtw_set_default_mgmt_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         int link_id,
+ #endif
+         u8 key_index)
+@@ -4908,7 +4908,7 @@ static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *nd
+ 	return ret;
+ }
+ 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0) || defined(BUILD_OPENWRT))
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev, unsigned int link_id)
+ #else
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
+@@ -9670,7 +9670,7 @@ void rtw_wdev_unregister(struct wireless_dev *wdev)
+ 	rtw_cfg80211_indicate_scan_done(adapter, _TRUE);
+ 
+ 	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)) || defined(COMPAT_KERNEL_RELEASE)
+-	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0))
++	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 0) || defined(BUILD_OPENWRT))
+ 	if (wdev->connected) {
+ 	#else
+ 	if (wdev->current_bss) {
+--- a/os_dep/linux/os_intfs.c
++++ b/os_dep/linux/os_intfs.c
+@@ -1618,11 +1618,7 @@ int rtw_os_ndev_register(_adapter *adapter, const char *name)
+ 	u8 rtnl_lock_needed = rtw_rtnl_lock_needed(dvobj);
+ 
+ #ifdef CONFIG_RTW_NAPI
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+-        netif_napi_add_weight(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
+-#else
+-        netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
+-#endif
++netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll);
+ #endif /* CONFIG_RTW_NAPI */
+ 
+ #if defined(CONFIG_IOCTL_CFG80211)
+ 
diff --git a/.github/diy/packages/rtl8189es/patches/020-wireless-5.8.patch b/.github/diy/packages/rtl8189es/patches/020-wireless-5.8.patch
new file mode 100644
index 000000000..bfdc2d28e
--- /dev/null
+++ b/.github/diy/packages/rtl8189es/patches/020-wireless-5.8.patch
@@ -0,0 +1,86 @@
+--- a/os_dep/linux/ioctl_cfg80211.c
++++ b/os_dep/linux/ioctl_cfg80211.c
+@@ -438,9 +438,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset, u8
+ 	ret = rtw_chbw_to_cfg80211_chan_def(wiphy, &chdef, ch, bw, offset, ht);
+ 	if (ret != _SUCCESS)
+ 		goto exit;
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,3, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,1, 0))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5,15, 2))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0);
+ #else
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef);
+@@ -1114,7 +1114,7 @@ void rtw_cfg80211_indicate_connect(_adapter *padapter)
+ 		#endif
+ 
+ #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 12, 0)
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0) || defined(BUILD_OPENWRT))
+ 		roam_info.links[0].bssid = cur_network->network.MacAddress;
+ #else
+ 		roam_info.bssid = cur_network->network.MacAddress;
+@@ -1681,7 +1681,7 @@ static int rtw_cfg80211_set_encryption(struct net_device *dev, struct ieee_param
+ }
+ 
+ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	,int link_id
+ #endif
+ 	, u8 key_index
+@@ -1828,7 +1828,7 @@ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+ }
+ 
+ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	,int link_id
+ #endif
+ 	, u8 keyid
+@@ -1996,7 +1996,7 @@ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+ }
+ 
+ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
+@@ -2020,7 +2020,7 @@ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+ 
+ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ 	struct net_device *ndev, 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif	 	
+ 	u8 key_index
+@@ -2072,7 +2072,7 @@ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 30))
+ int cfg80211_rtw_set_default_mgmt_key(struct wiphy *wiphy,
+ 	struct net_device *ndev, 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif	
+ 	u8 key_index)
+@@ -5167,7 +5167,7 @@ static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *nd
+ 
+ 	return ret;
+ }
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2) || defined(BUILD_OPENWRT))
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev, unsigned int link_id)
+ #else
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
+@@ -10270,7 +10270,7 @@ void rtw_wdev_unregister(struct wireless_dev *wdev)
+ 	rtw_cfg80211_indicate_scan_done(adapter, _TRUE);
+ 
+ 	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)) || defined(COMPAT_KERNEL_RELEASE)
+-	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2) || defined(BUILD_OPENWRT))
+ 	if (wdev->connected) {
+ 	#else
+ 	if (wdev->current_bss) {
diff --git a/.github/diy/packages/rtl8192eu/patches/030-wireless-5.8.patch b/.github/diy/packages/rtl8192eu/patches/030-wireless-5.8.patch
new file mode 100644
index 000000000..eb6f0ae53
--- /dev/null
+++ b/.github/diy/packages/rtl8192eu/patches/030-wireless-5.8.patch
@@ -0,0 +1,112 @@
+--- a/os_dep/linux/ioctl_cfg80211.c
++++ b/os_dep/linux/ioctl_cfg80211.c
+@@ -439,9 +439,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
+ 
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0))
+ 	if (started) {
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false);
+ #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, false);
+@@ -454,9 +454,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
+ 
+ 	if (!rtw_cfg80211_allow_ch_switch_notify(adapter))
+ 		goto exit;
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5,15, 2))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0);
+ #else
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef);
+@@ -1132,7 +1132,7 @@ check_bss:
+ 		#endif
+ 
+ 		#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 12, 0)
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0) || defined(BUILD_OPENWRT))
+ 		roam_info.links[0].bssid = cur_network->network.MacAddress;
+ #else
+ 		roam_info.bssid = cur_network->network.MacAddress;
+@@ -1698,7 +1698,7 @@ exit:
+ }
+ 
+ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	,int link_id
+ #endif
+ 	, u8 key_index
+@@ -1845,7 +1845,7 @@ addkey_end:
+ }
+ 
+ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	,int link_id
+ #endif
+ 	, u8 keyid
+@@ -2013,7 +2013,7 @@ exit:
+ }
+ 
+ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
+@@ -2037,7 +2037,7 @@ static int cfg80211_rtw_del_key(struct w
+ 
+ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ 	struct net_device *ndev, 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif	 	
+ 	u8 key_index
+@@ -2089,7 +2089,7 @@ static int cfg80211_rtw_set_default_key(
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 30))
+ int cfg80211_rtw_set_default_mgmt_key(struct wiphy *wiphy,
+ 	struct net_device *ndev, 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif	
+ 	u8 key_index)
+@@ -5102,7 +5102,7 @@ static int cfg80211_rtw_change_beacon(st
+ 	return ret;
+ }
+ 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2) || defined(BUILD_OPENWRT))
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev, unsigned int link_id)
+ #else
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
+@@ -10163,7 +10163,7 @@ void rtw_wdev_unregister(struct wireless
+ 	rtw_cfg80211_indicate_scan_done(adapter, _TRUE);
+ 
+ 	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)) || defined(COMPAT_KERNEL_RELEASE)
+-	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2) || defined(BUILD_OPENWRT))
+ 	if (wdev->connected) {
+ 	#else
+ 	if (wdev->current_bss) {
+--- a/os_dep/linux/os_intfs.c
++++ b/os_dep/linux/os_intfs.c
+@@ -1676,7 +1676,10 @@ int rtw_os_ndev_register(_adapter *adapt
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+ 	netif_napi_add_weight(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT
+ #else
+-	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT
++	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll
++#ifndef BUILD_OPENWRT
++		,RTL_NAPI_WEIGHT
++#endif
+ #endif
+ 	);
+ #endif /* CONFIG_RTW_NAPI */
diff --git a/.github/diy/packages/rtl8812au-ac/patches/040-wireless-5.8.patch b/.github/diy/packages/rtl8812au-ac/patches/040-wireless-5.8.patch
new file mode 100644
index 000000000..9f39affbe
--- /dev/null
+++ b/.github/diy/packages/rtl8812au-ac/patches/040-wireless-5.8.patch
@@ -0,0 +1,139 @@
+--- a/os_dep/linux/ioctl_cfg80211.c
++++ b/os_dep/linux/ioctl_cfg80211.c
+@@ -460,9 +460,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
+ 
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 19, 0))
+ 	if (started) {
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, 0, false);
+ #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 11, 0)) || (RHEL_RELEASE_CODE >= RHEL_RELEASE_VERSION(8,0))
+ 		cfg80211_ch_switch_started_notify(adapter->pnetdev, &chdef, 0, false);
+@@ -476,9 +476,9 @@ u8 rtw_cfg80211_ch_switch_notify(_adapter *adapter, u8 ch, u8 bw, u8 offset,
+ 	if (!rtw_cfg80211_allow_ch_switch_notify(adapter))
+ 		goto exit;
+ 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 3, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(6, 9, 0))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0, 0);
+-#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
++#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 15, 2))
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef, 0);
+ #else
+ 	cfg80211_ch_switch_notify(adapter->pnetdev, &chdef);
+@@ -1155,7 +1155,7 @@ check_bss:
+ 		#endif
+ 
+ 		#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 12, 0)
+-		#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 0, 0)
++		#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 0, 0) && !defined(BUILD_OPENWRT))
+ 		roam_info.channel = notify_channel;
+ 		roam_info.bssid = cur_network->network.MacAddress;
+ 		#endif
+@@ -1738,7 +1738,7 @@ exit:
+ }
+ 
+ static int cfg80211_rtw_add_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	, int link_id
+ #endif
+ 	, u8 key_index
+@@ -1885,7 +1885,7 @@ addkey_end:
+ }
+ 
+ static int cfg80211_rtw_get_key(struct wiphy *wiphy, struct net_device *ndev
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	, int link_id
+ #endif
+ 	, u8 keyid
+@@ -2053,7 +2053,7 @@ exit:
+ }
+ 
+ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
+@@ -2077,7 +2077,7 @@ static int cfg80211_rtw_del_key(struct w
+ 
+ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif
+ 	u8 key_index
+@@ -2129,7 +2129,7 @@ static int cfg80211_rtw_set_default_key(
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 30))
+ int cfg80211_rtw_set_default_mgmt_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+ 	int link_id,
+ #endif
+ 	u8 key_index)
+@@ -5315,7 +5315,7 @@ static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *nd
+ 	return ret;
+ }
+ 
+-#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 19, 2))
++#if (LINUX_VERSION_CODE < KERNEL_VERSION(5, 15, 2))
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev)
+ #else
+ static int cfg80211_rtw_stop_ap(struct wiphy *wiphy, struct net_device *ndev, unsigned int link_id)
+@@ -6143,7 +6143,7 @@ static int	cfg80211_rtw_set_channel(stru
+ 	return 0;
+ }
+ 
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2) || defined(BUILD_OPENWRT))
+ static int cfg80211_rtw_get_channel(struct wiphy *wiphy, struct wireless_dev *wdev, unsigned int link_id, struct cfg80211_chan_def *chandef){
+ #else
+ static int cfg80211_rtw_get_channel(struct wiphy *wiphy, struct wireless_dev *wdev, struct cfg80211_chan_def *chandef){
+@@ -10484,9 +10484,9 @@ void rtw_wdev_unregister(struct wireless_dev *wdev)
+ 	rtw_cfg80211_indicate_scan_done(adapter, _TRUE);
+ 
+ 	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(3, 11, 0)) || defined(COMPAT_KERNEL_RELEASE)
+-	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
++	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2) || defined(BUILD_OPENWRT))
+ 	if (wdev->links[0].client.current_bss) {
+-	#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2))
++	#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5, 19, 2) || defined(BUILD_OPENWRT))
+ 	if (wdev->connected) {
+ 	#else
+ 	if (wdev->current_bss) {
+--- a/os_dep/linux/os_intfs.c
++++ b/os_dep/linux/os_intfs.c
+@@ -1737,7 +1737,11 @@ int rtw_os_ndev_register(_adapter *adapt
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
+ 	netif_napi_add_weight(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
+ #else
+-	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll, RTL_NAPI_WEIGHT);
++	netif_napi_add(ndev, &adapter->napi, rtw_recv_napi_poll
++#ifndef BUILD_OPENWRT
++		, RTL_NAPI_WEIGHT
++#endif
++	);
+ #endif
+ #endif /* CONFIG_RTW_NAPI */
+ 
+@@ -2794,7 +2798,13 @@ static int netdev_vir_if_close(struct ne
+ 
+ #ifdef CONFIG_IOCTL_CFG80211
+ 	wdev->iftype = NL80211_IFTYPE_MONITOR;
++	#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 0, 0) || defined(BUILD_OPENWRT))
++	wdev->links[0].client.current_bss = NULL;
++	#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(5,19, 2))
++	wdev->connected = NULL;
++	#else
+ 	wdev->current_bss = NULL;
++	#endif
+ 	rtw_scan_abort(padapter);
+ 	rtw_cfg80211_wait_scan_req_empty(padapter, 200);
+ 	adapter_wdev_data(padapter)->bandroid_scan = _FALSE;
diff --git a/.github/diy/patches/advancedplus.patch b/.github/diy/patches/advancedplus.patch
index 982db3005..635deb4b4 100644
--- a/.github/diy/patches/advancedplus.patch
+++ b/.github/diy/patches/advancedplus.patch
@@ -1,55 +1,48 @@
 --- a/luci-app-advancedplus/luasrc/controller/advancedplus.lua
 +++ b/luci-app-advancedplus/luasrc/controller/advancedplus.lua
-@@ -9,18 +9,20 @@ function index()
- 	end
- 
- 	local page
--	page = entry({"admin","system","advancedplus"},alias("admin","system","advancedplus","kucatset"),_("Advanced plus"),61)
-+	page = entry({"admin","system","advancedplus"},alias("admin","system","advancedplus","advancededit"),_("Advanced plus"),61)
- 	page.dependent = true
- 	page.acl_depends = { "luci-app-advancedplus" }
--	entry({"admin","system","advancedplus","kucatset"},cbi("advancedplus/kucatset"),_("KuCat Theme Config"),20).leaf = true
--	if nixio.fs.access('/www/luci-static/argon/css/cascade.css') then
-+	if uci.cursor():get("luci", "main", "mediaurlbase") == "/luci-static/kucat" then
-+	    entry({"admin","system","advancedplus","kucatset"},cbi("advancedplus/kucatset"),_("KuCat Theme Config"),20).leaf = true
+@@ -24,6 +24,7 @@ function index()
+ 	entry({"admin","system","advancedplus","advancedipk"},cbi("advancedplus/advancedipk", {hideapplybtn=true, hidesavebtn=true, hideresetbtn=true}),_("Loading plugins"),30).leaf = true
+ 	if nixio.fs.access('/www/luci-static/kucat/css/style.css') then
+ 	    entry({"admin","system","advancedplus","kucatset"},cbi("advancedplus/kucatset"),_("KuCat Theme Config"),40).leaf = true
 +	    entry({"admin", "system","advancedplus","kucatupload"}, form("advancedplus/kucatupload"), _("Desktop background upload"), 80).leaf = true
-+	end
-+	if uci.cursor():get("luci", "main", "mediaurlbase") == "/luci-static/argon" then
- 	    entry({"admin", "system", "advancedplus","argon-config"}, form("advancedplus/argon-config"), _("Argon Config"), 30).leaf = true
  	end
--	if nixio.fs.access('/www/luci-static/design/css/style.css') then
-+	if uci.cursor():get("luci", "main", "mediaurlbase") == "/luci-static/design" then
- 	    entry({"admin", "system",  "advancedplus","design-config"}, form("advancedplus/design-config"), _("Design Config"), 40).leaf = true
+ 	if nixio.fs.access('/www/luci-static/argon/css/cascade.css') then
+ 	    entry({"admin", "system", "advancedplus","argon-config"}, form("advancedplus/argon-config"), _("Argon Config"), 50).leaf = true
+@@ -32,7 +33,6 @@ function index()
+ 	    entry({"admin", "system",  "advancedplus","design-config"}, form("advancedplus/design-config"), _("Design Config"), 60).leaf = true
  	end
- 	entry({"admin","system","advancedplus","advancedset"},cbi("advancedplus/advancedset"),_("Advanced Setting"),10).leaf = true
--	entry({"admin","system","advancedplus","advancededit"},cbi("advancedplus/advancededit"),_("Advanced Edit"),60).leaf = true
-+	entry({"admin","system","advancedplus","advancededit"},cbi("advancedplus/advancededit"),_("Advanced Edit"),1).leaf = true
  	entry({"admin", "system","advancedplus","upload"}, form("advancedplus/upload"), _("Login Background Upload"), 70).leaf = true
 -	entry({"admin", "system","advancedplus","kucatupload"}, form("advancedplus/kucatupload"), _("Desktop background upload"), 80).leaf = true
- end
+ 	
+ 
+ 	entry({"admin", "system", "advancedplus", "advancedrun"}, call("advanced_run"))
 
 --- a/luci-app-advancedplus/luasrc/model/cbi/advancedplus/advancedset.lua
 +++ b/luci-app-advancedplus/luasrc/model/cbi/advancedplus/advancedset.lua
-@@ -24,7 +24,9 @@ ul:depends("qos", true)
+@@ -24,11 +24,9 @@ ul:depends("qos", true)
  
- -- e = t:option(Flag, "uhttps",translate('Accessing using HTTPS'), translate('Open the address in the background and use HTTPS for secure access'))
+ e = t:option(Flag, "uhttps",translate('Accessing using HTTPS'), translate('Open the address in the background and use HTTPS for secure access'))
  
 +if fs.access('/usr/bin/zsh') then
  e = t:option(Flag, "usshmenu",translate('No backend menu required'), translate('OPENWRT backend and SSH login do not display shortcut menus'))
+-
+-e = t:option(Flag, "wizard",translate('Hide Wizard'), translate('Show or hide the setup wizard menu'))
+-e.default = "0"
+-e.rmempty = false
 +end
  
- if fs.access('/etc/config/netwizard') then
- e = t:option(Flag, "wizard",translate('Hide Wizard'), translate('Show or hide the setup wizard menu'))
+ e = t:option(Flag, "tsoset",translate('TSO optimization for network card interruption'), translate('Turn off the TSO parameters of the INTEL225 network card to improve network interruption'))
+ e.default = "1"
 
 --- a/luci-app-advancedplus/root/etc/init.d/advancedplus
 +++ b/luci-app-advancedplus/root/etc/init.d/advancedplus
-@@ -144,8 +144,10 @@ dev=`ifconfig | grep "Point-to-Point" | cut -d " " -f1`
+@@ -233,8 +233,10 @@ tsoset
      fi
      #cpumode=`uci -q get advancedplus.@basic[0].cpumode`
      #[ ! $cpumode ] || cpumodeset $cpumode   /usr/bin/zsh
-+if [ "$(which zsh)" ]; then
++    if [ "$(which zsh)" ]; then
      sed -i "\/bin\/zsh/d" /etc/profile
      [ "x$(uci -q get advancedplus.@basic[0].usshmenu)" = "x1" ] || echo '/usr/bin/zsh'  >> /etc/profile
-+fi
++    fi
      uci commit netwizard 
  }
diff --git a/.github/diy/patches/luci-app-filebrowser.patch b/.github/diy/patches/luci-app-filebrowser.patch
index a527c9b8b..3d845f670 100644
--- a/.github/diy/patches/luci-app-filebrowser.patch
+++ b/.github/diy/patches/luci-app-filebrowser.patch
@@ -1,5 +1,5 @@
---- a/luci-app-filebrowser/htdocs/luci-static/resources/view/filebrowser.js
-+++ b/luci-app-filebrowser/htdocs/luci-static/resources/view/filebrowser.js
+--- a/luci-app-filebrowser-go/htdocs/luci-static/resources/view/filebrowser.js
++++ b/luci-app-filebrowser-go/htdocs/luci-static/resources/view/filebrowser.js
 @@ -28,7 +28,7 @@ function renderStatus(isRunning, port) {
  	if (isRunning) {
  		var button = String.format('&#160;<a class="btn cbi-button" href="http://%s:%s" target="_blank" rel="noreferrer noopener">%s</a>',
diff --git a/.github/diy/patches/luci-app-turboacc.patch b/.github/diy/patches/luci-app-turboacc.patch
index 5a2d6634d..52c49dd27 100644
--- a/.github/diy/patches/luci-app-turboacc.patch
+++ b/.github/diy/patches/luci-app-turboacc.patch
@@ -102,7 +102,7 @@
  	option hw_flow '1'
  	option sfe_flow '1'
 -	option fullcone_nat '2'
-+	option fullcone_nat '1'
++	option fullcone_nat '0'
  	option hw_wed '0'
  	option bbr_cca '0'
  	option dns_caching '0'
diff --git a/.github/diy/patches/miniupnpd.patch b/.github/diy/patches/miniupnpd.patch
new file mode 100644
index 000000000..c090e68de
--- /dev/null
+++ b/.github/diy/patches/miniupnpd.patch
@@ -0,0 +1,13 @@
+--- a/miniupnpd/files/upnpd.config
++++ b/miniupnpd/files/upnpd.config
+@@ -12,6 +12,10 @@ config upnpd config
+ 	option port		5000
+ 	option upnp_lease_file	/var/run/miniupnpd.leases
+ 	option igdv1		1
++	option ipv6_disable 1
++	option stun_host 'stun.qq.com'
++	option stun_port '3478'
++	option use_stun '1'
+ 
+ config perm_rule
+ 	option action		allow
diff --git a/.github/workflows/upstream.yml b/.github/workflows/upstream.yml
index 3957a11c7..50cf76910 100644
--- a/.github/workflows/upstream.yml
+++ b/.github/workflows/upstream.yml
@@ -94,7 +94,6 @@ jobs:
         git_clone https://github.com/kiddin9/openwrt-bypass && mvdir openwrt-bypass
         git_clone https://github.com/kiddin9/luci-app-dnsfilter
         git_clone https://github.com/kiddin9/aria2
-        git_clone https://github.com/kiddin9/luci-app-baidupcs-web
         git_clone https://github.com/kiddin9/luci-theme-edge
         git_clone https://github.com/kiddin9/autoshare && mvdir autoshare
         git_clone https://github.com/kiddin9/luci-app-xlnetacc
@@ -173,7 +172,6 @@ jobs:
         ) &
         (
         git_clone https://github.com/tty228/luci-app-wechatpush
-        git_clone https://github.com/4IceG/luci-app-sms-tool smstool && mvdir smstool
         git_clone https://github.com/4IceG/luci-app-modemband modemb  && mv -n modemb/* ./;rm -rf modemb
         git_clone https://github.com/ZeaKyX/speedtest-web
         git_clone https://github.com/ZeaKyX/luci-app-speedtest-web
@@ -201,7 +199,7 @@ jobs:
         ) &
         (
         git_clone https://github.com/linkease/openwrt-app-actions
-        git_clone https://github.com/xiaorouji/openwrt-passwall-packages && rm -rf openwrt-passwall-packages/{naiveproxy,chinadns-ng} && mvdir openwrt-passwall-packages
+        git_clone https://github.com/xiaorouji/openwrt-passwall-packages && rm -rf openwrt-passwall-packages/chinadns-ng && mvdir openwrt-passwall-packages
         ) &
         (
         git_clone https://github.com/honwen/luci-app-shadowsocks-rust
@@ -230,6 +228,7 @@ jobs:
         git_clone https://github.com/sirpdboy/luci-app-chatgpt-web
         git_clone https://github.com/sirpdboy/luci-app-eqosplus && rm -rf luci-app-eqosplus/po/zh_Hans
         git_clone https://github.com/JiaY-shi/fancontrol fanc && mvdir fanc
+        git_clone https://github.com/Siriling/5G-Modem-Support Modem && mvdir Modem;rm -rf rooter
         ) &
         (
         svn export https://github.com/QiuSimons/OpenWrt-Add/trunk/luci-app-irqbalance
@@ -239,13 +238,12 @@ jobs:
         svn export https://github.com/HiGarfield/lede-17.01.4-Mod/trunk/package/extra/luci-app-openvpn-server
         ) &
         (
-        git_sparse_clone master "https://github.com/coolsnowwolf/packages" "leanpkg" net/miniupnpd net/mwan3 multimedia/UnblockNeteaseMusic-Go \
-        multimedia/UnblockNeteaseMusic net/amule net/softethervpn5 net/antileech net/baidupcs-web multimedia/gmediarender net/go-aliyundrive-webdav \
+        git_sparse_clone master "https://github.com/coolsnowwolf/packages" "leanpkg" multimedia/UnblockNeteaseMusic-Go \
+        multimedia/UnblockNeteaseMusic net/amule net/softethervpn5 net/antileech multimedia/gmediarender net/go-aliyundrive-webdav \
         net/qBittorrent-static net/phtunnel net/frp \
-        net/uugamebooster net/verysync net/vlmcsd net/dnsforwarder net/tcpping net/netatalk net/pgyvpn
-        mv -f miniupnpd miniupnpd-iptables
+        net/uugamebooster net/verysync net/vlmcsd net/dnsforwarder net/tcpping net/netatalk net/pgyvpn lang/golang
         git_sparse_clone openwrt-23.05 "https://github.com/openwrt/packages" "oppackages" \
-        utils/coremark utils/watchcat utils/dockerd utils/cgroupfs-mount net/uwsgi net/ddns-scripts net/smartdns net/curl \
+        utils/coremark utils/watchcat net/miniupnpd net/snort3 utils/dockerd utils/cgroupfs-mount net/uwsgi net/ddns-scripts net/smartdns net/curl \
         net/ariang net/rp-pppoe
         ) &
         (
@@ -261,14 +259,14 @@ jobs:
 
         (
         git_sparse_clone openwrt-23.05 "https://github.com/immortalwrt/packages" "immpkgs" net/mwol net/n2n net/dae \
-        net/sub-web net/dnsproxy net/haproxy net/v2raya net/cdnspeedtest net/keepalived net/microsocks \
-        net/go-nats net/go-wol net/simple-torrent net/bitsrunlogin-go net/transfer net/cloudreve net/daed \
+        net/sub-web net/dnsproxy net/haproxy net/v2raya net/cdnspeedtest net/keepalived \
+        net/go-nats net/go-wol net/bitsrunlogin-go net/transfer net/cloudreve net/daed \
         net/subconverter net/ngrokc net/oscam net/njitclient net/scutclient net/gost net/ua2f net/dufs net/qBittorrent-Enhanced-Edition \
         net/tinyportmapper net/tinyfecvpn net/nexttrace net/pcap-dnsproxy net/rustdesk-server net/tuic-server net/speedtest-go \
-        net/dns-forwarder net/ipset-lists net/ShadowVPN net/adguardhome net/cloudflared net/nps net/naiveproxy \
-        libs/jpcre2 libs/wxbase libs/rapidjson libs/libcron libs/quickjspp libs/toml11 libs/libtorrent-rasterbar \
-        libs/libdouble-conversion libs/qt6base libs/cxxopts libs/jpcre2 libs/alac sound/spotifyd \
-        utils/qt6tools utils/cpulimit utils/filebrowser utils/cups utils/cups-bjnp utils/joker net/udp2raw net/msd_lite \
+        net/dns-forwarder net/ipset-lists net/ShadowVPN net/adguardhome net/cloudflared net/nps \
+        libs/wxbase libs/rapidjson libs/libcron libs/quickjspp libs/toml11 libs/libtorrent-rasterbar \
+        libs/libdouble-conversion libs/qt6base libs/cxxopts libs/jpcre2 libs/alac libs/libcryptopp sound/spotifyd \
+        utils/qt6tools utils/cpulimit utils/sendat utils/exfatprogs utils/filebrowser utils/cups utils/cups-bjnp utils/joker net/udp2raw net/msd_lite \
         multimedia/you-get multimedia/lux multimedia/ykdl multimedia/gallery-dl devel/go-rice admin/gotop admin/netdata
         ) &
         (
@@ -285,17 +283,17 @@ jobs:
         ) &
         git_sparse_clone master "https://github.com/coolsnowwolf/luci" "leluci" applications libs/luci-lib-fs
         mv -f applications luciapp
-        rm -rf luciapp/{luci-app-qbittorrent,luci-app-cpufreq,luci-app-e2guardian,luci-app-aliyundrive-fuse}
+        rm -rf luciapp/{luci-app-qbittorrent,luci-app-cpufreq,luci-app-e2guardian,luci-app-aliyundrive-fuse,luci-app-syncdial}
         git_sparse_clone master "https://github.com/immortalwrt/luci" "immluci" applications protocols/luci-proto-minieap
         mv -n applications/* luciapp/; rm -rf applications
 
-        for ipk in $(ls -d luciapp/!(luci-app-rclone|luci-app-mwan3)/); do
+        for ipk in $(ls -d luciapp/!(luci-app-rclone)/); do
         	if [[ $(ls $ipk/po | wc -l) -gt 3 ]]; then
             rm -rf $ipk
         	fi
         done
 
-        git_sparse_clone master "https://github.com/coolsnowwolf/lede" "leanlede" package/lean package/qca/shortcut-fe package/wwan package/network/services/shellsync
+        git_sparse_clone master "https://github.com/coolsnowwolf/lede" "leanlede" package/lean package/qca/shortcut-fe package/network/services/shellsync
 
     - name: Delete duplicated packages
       run: |
@@ -303,7 +301,6 @@ jobs:
         mv -n luciapp/!(luci-app-noddos|luci-app-cshark|luci-app-dnscrypt-proxy|luci-app-https-dns-proxy|luci-app-ssr-mudb-server|luci-app-ledtrig-*) ./ ; rm -Rf luciapp
         mv -n lean/* ./ ; rm -Rf lean
         mv -n liep/* ./ ; rm -Rf liep
-        mv -n wwan/*/* ./; rm -rf wwan
         mv -n openwrt-app-actions/applications/* ./;rm -rf openwrt-app-actions
         rm -Rf */.git
         for pkg in $(ls -d .github/diy/packages/*); do
@@ -336,7 +333,7 @@ jobs:
              -H "Authorization: Bearer ${{ secrets.TOKEN_KIDDIN9 }}" \
              -X POST -d '{ "query": "query {repository(owner: \"'"$1"'\", name: \"'"$2"'\"){latestRelease{tagName tagCommit{oid}}refs(refPrefix:\"refs/tags/\",last:1,orderBy:{field:TAG_COMMIT_DATE,direction:ASC}){edges{node{name target{oid}}}}defaultBranchRef{target{...on Commit {oid}}}}}"}' https://api.github.com/graphql)
         }
-        for pkg in $(ls !(luci-*|cloudflared|joker|aria2|frp|chinadns-ng|ddns-go|libtorrent-rasterbar|ua2f|homebox|xtables-wgobfs|nexttrace|tuic-server|subconverter|softethervpn5|shadowsocksr-libev|mt76|sub-web|mmdvm-host|alist|curl|qBittorrent-Enhanced-Edition|libwxwidgets|ysf-clients|natter|fullconenat|v2raya|r8152|amule|brook|3proxy|containerd|oscam|netmaker|n2n_v2|naiveproxy|miniupnpd-nft|upx|netdata|tailscale|mosdns|r8168|coremark|mbedtls|rtl8821cu|rtl8188eu|rtl8812au-ac|rtl88x2bu|rtl8189es|quickjspp|accel-ppp)/Makefile); do
+        for pkg in $(ls !(luci-*|cloudflared|joker|aria2|frp|libcryptopp|toml11|chinadns-ng|ddns-go|libtorrent-rasterbar|ua2f|homebox|xtables-wgobfs|nexttrace|tuic-server|subconverter|softethervpn5|shadowsocksr-libev|mt76|sub-web|mmdvm-host|alist|curl|qBittorrent-Enhanced-Edition|libwxwidgets|ysf-clients|natter|fullconenat|v2raya|r8152|amule|brook|3proxy|containerd|oscam|netmaker|n2n_v2|naiveproxy|miniupnpd-nft|upx|netdata|tailscale|mosdns|r8168|coremark|mbedtls|quickjspp|accel-ppp)/Makefile); do
           repo="$(grep PKG_SOURCE_URL $pkg | grep github | cut -f 4-5 -d '/' | sed -e 's/.git//' | grep -E '[0-9a-zA-Z_-]+$')" || true
           if [ "$repo" ]; then
             owner="$(echo $repo | cut -f 1 -d '/')"
@@ -390,17 +387,11 @@ jobs:
         bash .github/diy/convert_translation.sh -a >/dev/null 2>&1 || true
         
         #sed -i 's?"http://" + window.location.hostname?window.location.protocol + "//" + window.location.hostname?g' `find luci-app*/luasrc/view/ -maxdepth 2 -name "*.htm"`
-        rm -f miniupnpd-iptables/patches/106-spam-syslog*
-        sed -i "/minisign:minisign/d" luci-app-dnscrypt-proxy2/Makefile
-        sed -i 's/+libstdcpp/+libstdcpp +zlib/' ngrokc/Makefile
         sed -i 's/+rclone\( \|$\)/+rclone +fuse-utils\1/g' luci-app-rclone/Makefile
-        sed -i 's/+libcap /+libcap +libcap-bin /' luci-app-openclash/Makefile
         sed -i 's/\(+luci-compat\)/\1 +luci-theme-argon/' luci-app-argon-config/Makefile
-        sed -i 's/+vsftpd$/+vsftpd-alt/' luci-app-vsftpd/Makefile
         sed -i 's/ +uhttpd-mod-ubus//' luci-app-packet-capture/Makefile
         sed -i '/boot()/,+2d' ddns-scripts/files/etc/init.d/ddns
         sed -i "/DISTRIB_DESCRIPTION/c\DISTRIB_DESCRIPTION=\"%D %C by Kiddin'\"" base-files/files/etc/openwrt_release
-        sed -i "s/PKG_VERSION:=2/PKG_VERSION:=3/" mwan3/Makefile
         sed -i "s/default n/default y/" dockerd/Config.in
         sed -i "/+uhttpd/d" ariang/Makefile
         sed -i -e '$a /etc/bench.log' \
@@ -410,14 +401,11 @@ jobs:
         sed -i -e '/^\/etc\/profile/d' \
         -e '/^\/etc\/shinit/d' \
         base-files/Makefile
-        cp -f base-files/files/sbin/sysupgrade my-default-settings/files/sbin/
         sed -i '$a cgi-timeout = 300' uwsgi/files-luci-support/luci-webui.ini
         sed -i '$a cgi-timeout = 90' uwsgi/files-luci-support/luci-cgi_io.ini
         sed -i '/limit-as/c\limit-as = 5000' uwsgi/files-luci-support/luci-webui.ini
         sed -i 's/procd_set_param stderr 1/procd_set_param stderr 0/' uwsgi/files/uwsgi.init
-        sed -i 's/	ip.neighbors/	luci.ip.neighbors/' luci-app-wifidog/luasrc/model/cbi/wifidog/wifidog_cfg.lua
         sed -i 's/ if aarch64||arm||i386||x86_64//' luci-app-ssr-plus/Makefile
-        sed -i 's/transmission-daemon$/transmission-daemon +transmission-web-control/' luci-app-transmission/Makefile
         mkdir -p luci-app-passwall/root/www/luci-static/passwall
         cp -rf luci-app-bypass/root/www/luci-static/bypass/* luci-app-passwall/root/www/luci-static/passwall/
         cp -rf luci-app-bypass/root/usr/share/bypass/GeoLite2-Country.mmdb luci-app-passwall/root/usr/share/passwall/
@@ -431,6 +419,7 @@ jobs:
           -e 's/+ca-certificates/+ca-bundle/' \
           -e 's/php7/php8/g' \
           -e 's/+docker /+docker +dockerd /g' \
+          -e 's/URL:=@IMMORTALWRT/URL:=https:\/\/mirror2.immortalwrt.org\/sources/g' \
           */Makefile
         ) &
         
diff --git a/.gitignore b/.gitignore
index 01665bd80..67043824d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -16,6 +16,7 @@
 /luci-app-serverchan
 /luci-app-eqos
 /luci-app-webdav
+/luci-app-baidupcs-web
 /mentohust
 /subweb
 /wget
diff --git a/miniupnpd/files/upnpd.config b/miniupnpd/files/upnpd.config
index bd7c3ec40..31e6cce1f 100644
--- a/miniupnpd/files/upnpd.config
+++ b/miniupnpd/files/upnpd.config
@@ -12,6 +12,10 @@ config upnpd config
 	option port		5000
 	option upnp_lease_file	/var/run/miniupnpd.leases
 	option igdv1		1
+	option ipv6_disable 1
+	option stun_host 'stun.qq.com'
+	option stun_port '3478'
+	option use_stun '1'
 
 config perm_rule
 	option action		allow
diff --git a/rtl8188eu/patches/030-wireless-5.8.patch b/rtl8188eu/patches/030-wireless-5.8.patch
index 60586a2f0..8556f682d 100644
--- a/rtl8188eu/patches/030-wireless-5.8.patch
+++ b/rtl8188eu/patches/030-wireless-5.8.patch
@@ -59,6 +59,24 @@
          int link_id,
  #endif
  #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 37)) || defined(COMPAT_KERNEL_RELEASE)
+@@ -2006,7 +2006,7 @@ static int cfg80211_rtw_del_key(struct wiphy *wiphy, struct net_device *ndev,
+
+ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         int link_id,
+ #endif
+         u8 key_index
+@@ -2058,7 +2058,7 @@ static int cfg80211_rtw_set_default_key(struct wiphy *wiphy,
+ #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 30))
+ int cfg80211_rtw_set_default_mgmt_key(struct wiphy *wiphy,
+ 	struct net_device *ndev,
+-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0))
++#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 1, 0) || defined(BUILD_OPENWRT))
+         int link_id,
+ #endif
+         u8 key_index)
 @@ -4908,7 +4908,7 @@ static int cfg80211_rtw_change_beacon(struct wiphy *wiphy, struct net_device *nd
  	return ret;
  }
